File: backend/api/views/films.py
from rest_framework import filters, generics, viewsets, permissions
from rest_framework.response import Response
from rest_framework.views import APIView
from rest_framework.decorators import action
import random
from django.conf import settings
from django.contrib.auth import get_user_model
from django.db.models import Case, When, IntegerField, Value, Q
from django_filters.rest_framework import DjangoFilterBackend
from django.shortcuts import get_object_or_404

from gallery.models import (Film, Genre, Country, Type, UserTopFilm)
from api.filters import FilmFilter
from api.serializers.films import (FilmDetailSerializer,
                                   SearchListFilmSerilizer,
                                   GenreSerializer,
                                   CountrySerializer,
                                   TypeSerializer,
                                   UserTopFilmSerializer,
                                   TopFilmSerializer)
from api.pagination import FilmSearchPagination
from talk_about.constants import (MIN_RATING,
                                  EXCLUDED_GENRES,
                                  MIN_SEARCH_VOTES,
                                  SEARCH_SUGGESTIONS_LIMIT)


User = get_user_model()


class FilmViewSet(viewsets.ModelViewSet):
    """Вьюсет для фильмов."""

    queryset = Film.objects.all().select_related('type').prefetch_related(
        'genres', 'countries', 'persons'
    ).distinct()
    serializer_class = FilmDetailSerializer
    permission_classes = []
    filter_backends = [
        DjangoFilterBackend, filters.SearchFilter, filters.OrderingFilter
    ]
    filterset_class = FilmFilter
    pagination_class = FilmSearchPagination

    search_fields = [
        'name',
        'alternative_name',
        'en_name'
    ]
    ordering_fields = [
        'year',
        'kinopoisk_rating',
        'imdb_rating',
        'movie_length'
    ]
    ordering = ['-kinopoisk_rating', '-year']

    def get_serializer_class(self):
        if self.action in ['list',
                           'random_top_films',
                           'discover',
                           'search_suggestions',
                           'search']:
            return SearchListFilmSerilizer
        return super().get_serializer_class()

    def get_random_films_base_queryset(self, min_rating=MIN_RATING):
        """
        Базовый queryset для случайных подборок:
        - рейтинг >= min_rating
        - есть постер
        - есть хотя бы один жанр
        - исключены нежелательные жанры
        """
        queryset = (
            Film.objects.filter(
                kinopoisk_rating__gte=min_rating,
                genres__isnull=False,  # 👈 есть хотя бы один жанр
            )
            .exclude(
                Q(poster_url__isnull=True) | Q(poster_url='')
            )
            .exclude(
                genres__name__in=EXCLUDED_GENRES
            )
            .select_related('type')
            .prefetch_related('genres', 'persons')
            .distinct()
        )

        return queryset

    @action(
        detail=False,
        methods=['get'],
        url_path='discover',
        url_name='discover'
    )
    def discover(self, request):
        """
        Умная рекомендация: случайные фильмы с высоким рейтингом и постерами,
        без короткометражек, концертов и документальных.
        """
        # min_rating = float(request.query_params.get('min_rating', MIN_RATING))
        queryset = self.get_random_films_base_queryset(min_rating=MIN_RATING)

        genres = request.query_params.get('genres')
        if genres:
            genre_ids = [int(g) for g in genres.split(',') if g.isdigit()]
            queryset = queryset.filter(genres__id__in=genre_ids).distinct()

        year_min = request.query_params.get('year_min')
        if year_min and year_min.isdigit():
            queryset = queryset.filter(year__gte=int(year_min))

        year_max = request.query_params.get('year_max')
        if year_max and year_max.isdigit():
            queryset = queryset.filter(year__lte=int(year_max))

        total = queryset.count()
        count = min(int(request.query_params.get('count', 50)), 200)

        if total == 0:
            films = []
        elif total <= count:
            films = list(queryset)
            random.shuffle(films)
        else:
            if settings.DATABASES['default']['ENGINE'] == 'django.db.backends.postgresql':
                films = queryset.order_by('?')[:count]
            else:
                pks = list(queryset.values_list('id', flat=True))
                random_pks = random.sample(pks, count)
                films = list(queryset.filter(id__in=random_pks))

        serializer = self.get_serializer(films, many=True)

        return Response({
            'count': len(films),
            'total_matching': total,
            'excluded_genres': EXCLUDED_GENRES,
            'results': serializer.data
        })

    @action(
        detail=False,
        methods=['get'],
        url_path='search-suggestions',
        url_name='search-suggestions',
        permission_classes=[permissions.AllowAny],
    )
    def search_suggestions(self, request):
        query = request.query_params.get('q', '').strip()

        # Не обрабатываем запрос с 1 буквой
        if len(query) < 2:
            return Response({
                'total': 0,
                'results': []
            })

        queryset = Film.objects.filter(
            Q(name__icontains=query) |
            Q(alternative_name__icontains=query) |
            Q(en_name__icontains=query),
        ).exclude(
            Q(name__isnull=True) &
            Q(alternative_name__isnull=True) &
            Q(en_name__isnull=True)
        ).select_related(
            'type'
        ).prefetch_related(
            'genres'
        ).annotate(
            # 🔥 приоритет популярным фильмам
            popularity_priority=Case(
                When(kinopoisk_votes__gte=MIN_SEARCH_VOTES, then=Value(1)),
                default=Value(0),
                output_field=IntegerField(),
            )
        ).order_by(
            '-popularity_priority',   # сначала популярные
            '-kinopoisk_rating',      # потом рейтинг
            '-kinopoisk_votes',       # потом голоса
            '-year'
        ).distinct()

        total = queryset.count()
        films = queryset[:SEARCH_SUGGESTIONS_LIMIT]

        serializer = SearchListFilmSerilizer(films, many=True)

        return Response({
            'total': total,
            'results': serializer.data
        })

    @action(
        detail=False,
        methods=['get'],
        url_path='search',
        url_name='search',
        permission_classes=[permissions.AllowAny],
    )
    def search(self, request):
        query = request.query_params.get('q', '').strip()

        if len(query) < 2:
            return Response({
                'count': 0,
                'next': None,
                'previous': None,
                'results': []
            })

        queryset = Film.objects.filter(
            Q(name__icontains=query) |
            Q(alternative_name__icontains=query) |
            Q(en_name__icontains=query)
        ).select_related(
            'type'
        ).prefetch_related(
            'genres'
        ).order_by(
            '-kinopoisk_rating',
            '-kinopoisk_votes',
            '-year'
        ).distinct()

        paginator = FilmSearchPagination()
        page = paginator.paginate_queryset(queryset, request)
        serializer = SearchListFilmSerilizer(page, many=True)

        return paginator.get_paginated_response(serializer.data)


class TypeList(generics.ListCreateAPIView):
    queryset = Type.objects.all()
    serializer_class = TypeSerializer
    permission_classes = [permissions.AllowAny]


class GenreList(generics.ListCreateAPIView):
    queryset = Genre.objects.all()
    serializer_class = GenreSerializer
    permission_classes = [permissions.AllowAny]


class CountryList(generics.ListCreateAPIView):
    queryset = Country.objects.all()
    serializer_class = CountrySerializer
    permission_classes = [permissions.AllowAny]


class MyTopFilmsView(APIView):
    permission_classes = [permissions.IsAuthenticated]

    def get(self, request):
        top_films = UserTopFilm.objects.filter(
            user=request.user
        ).select_related('film')

        serializer = UserTopFilmSerializer(top_films, many=True)
        return Response(serializer.data)

    def put(self, request):
        film_ids = request.data.get('film_ids', [])

        if len(film_ids) > 10:
            return Response(
                {'detail': 'В топе может быть максимум 10 фильмов.'},
                status=400
            )

        if len(film_ids) != len(set(film_ids)):
            return Response(
                {'detail': 'Фильм нельзя добавить в топ дважды.'},
                status=400
            )

        films = Film.objects.filter(id__in=film_ids)

        if films.count() != len(film_ids):
            return Response(
                {'detail': 'Один или несколько фильмов не найдены.'},
                status=400
            )

        films_by_id = {film.pk: film for film in films}

        UserTopFilm.objects.filter(user=request.user).delete()

        UserTopFilm.objects.bulk_create([
            UserTopFilm(
                user=request.user,
                film=films_by_id[film_id],
                position=index + 1
            )
            for index, film_id in enumerate(film_ids)
        ])

        top_films = UserTopFilm.objects.filter(
            user=request.user
        ).select_related('film')

        serializer = UserTopFilmSerializer(top_films, many=True)
        return Response(serializer.data)


class UserTopFilmsView(APIView):
    permission_classes = [permissions.AllowAny]

    def get(self, request, user_id):
        user = get_object_or_404(User, id=user_id)

        top_films = UserTopFilm.objects.filter(
            user=user
        ).select_related('film')

        serializer = UserTopFilmSerializer(top_films, many=True)
        return Response(serializer.data)

File: frontend/src/api/films.js
import api from './config'

const BASE_URL = 'http://localhost:8000'


/**
 * 🔍 Быстрый поиск (для dropdown)
 * Возвращает:
 * {
 *   total: number,
 *   results: Film[]
 * }
 */
export const searchFilms = async (query) => {
  try {
    const response = await api.get('/films/search-suggestions/', {
      params: { q: query }
    })

    return {
      total: response.data?.total || 0,
      results: response.data?.results || []
    }
  } catch (error) {
    console.error('Search suggestions error:', error)
    return {
      total: 0,
      results: []
    }
  }
}


/**
 * 🔎 Полный поиск (страница результатов)
 * Возвращает стандартную DRF пагинацию:
 * {
 *   count,
 *   next,
 *   previous,
 *   results
 * }
 */
export const searchFilmsFull = async (query, page = 1) => {
  try {
    const response = await api.get('/films/search/', {
      params: {
        q: query,
        page
      }
    })

    return response.data
  } catch (error) {
    console.error('Full search error:', error)
    return {
      count: 0,
      next: null,
      previous: null,
      results: []
    }
  }
}


/**
 * 🎬 Детали фильма
 */
export const getFilmDetails = async (id) => {
  try {
    const response = await api.get(`/films/${id}/`)
    return response.data
  } catch (error) {
    console.error('Film details error:', error)
    throw error
  }
}


/**
 * 🎲 Случайные топ-фильмы (главная)
 */
export const getRandomTopFilms = async (params = {}) => {
  try {
    const response = await api.get('/films/discover/', { params })
    return response.data
  } catch (error) {
    console.error('Random films error:', error)
    return {
      count: 0,
      results: []
    }
  }
}

export const getFilmGenres = async () => {
  try {
    const response = await api.get('/genres/')
    return response.data || []
  } catch (error) {
    console.error('Genres fetch error:', error)
    return []
  }
}

export const getFilmCountries = async () => {
  try {
    const response = await api.get('/countries/')
    return response.data || []
  } catch (error) {
    console.error('Countries fetch error:', error)
    return []
  }
}

export const getPickedFilms = async (filters = {}, page = 1) => {
  try {
    const params = new URLSearchParams()

    if (filters.year_min) params.append('year_min', filters.year_min)
    if (filters.year_max) params.append('year_max', filters.year_max)

    if (filters.movie_length_min) params.append('movie_length_min', filters.movie_length_min)
    if (filters.movie_length_max) params.append('movie_length_max', filters.movie_length_max)

    if (filters.kinopoisk_rating_min) params.append('kinopoisk_rating_min', filters.kinopoisk_rating_min)
    if (filters.kinopoisk_rating_max) params.append('kinopoisk_rating_max', filters.kinopoisk_rating_max)

    if (filters.imdb_rating_min) params.append('imdb_rating_min', filters.imdb_rating_min)
    if (filters.imdb_rating_max) params.append('imdb_rating_max', filters.imdb_rating_max)

    if (filters.age_rating) params.append('age_rating', filters.age_rating)

    if (Array.isArray(filters.genres)) {
      filters.genres.forEach((id) => params.append('genres', id))
    }

    if (Array.isArray(filters.countries)) {
      filters.countries.forEach((id) => params.append('countries', id))
    }

    if (Array.isArray(filters.persons)) {
      filters.persons.forEach((id) => params.append('persons', id))
    }

    params.append('page', page)

    const response = await api.get(`/films/?${params.toString()}`)
    return response.data
  } catch (error) {
    console.error('Picked films fetch error:', error)
    return {
      count: 0,
      next: null,
      previous: null,
      results: []
    }
  }
}